WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift

```diff
@@ -0,0 +1,221 @@
+import Foundation
+import Yosemite
+import protocol Storage.StorageManagerType
+import protocol WooFoundation.Analytics
+
+/// View model for `MostActiveCouponsCard`.
+///
+final class MostActiveCouponsCardViewModel: ObservableObject {
+    // Set externally to trigger callback upon hiding the Coupons card.
+    var onDismiss: (() -> Void)?
+
+    @Published private(set) var timeRange = StatsTimeRangeV4.today
+    @Published private(set) var syncingData = false
+    @Published private(set) var syncingError: Error?
+    @Published private(set) var coupons: [Coupon] = []
+    @Published var selectedCoupon: Coupon?
+
+    let siteID: Int64
+    let siteTimezone: TimeZone
+    private let stores: StoresManager
+    private let storage: StorageManagerType
+    private let analytics: Analytics
+
+    init(siteID: Int64,
+         siteTimezone: TimeZone = .siteTimezone,
+         stores: StoresManager = ServiceLocator.stores,
+         storage: StorageManagerType = ServiceLocator.storageManager,
+         analytics: Analytics = ServiceLocator.analytics) {
+        self.siteID = siteID
+        self.siteTimezone = siteTimezone
+        self.stores = stores
+        self.storage = storage
+        self.analytics = analytics
+
+        Task { @MainActor in
+            self.timeRange = await loadLastTimeRange() ?? .today
+            await reloadData()
+        }
+    }
+
+    func dismiss() {
+        // TODO: add tracking
+        onDismiss?()
+    }
+
+    func didSelectTimeRange(_ newTimeRange: StatsTimeRangeV4) {
+        timeRange = newTimeRange
+        saveLastTimeRange(timeRange)
+
+        Task { [weak self] in
+            await self?.reloadData()
+        }
+    }
+
+    @MainActor
+    func reloadData() async {
+        syncingData = true
+        syncingError = nil
+        do {
+            let activeCouponReports = try await mostActiveCoupons()
+            coupons = try await loadCouponDetails(for: activeCouponReports)
+        } catch {
+            syncingError = error
+            DDLogError("⛔️ Dashboard (Most active coupons) — Error loading most active coupons: \(error)")
+        }
+        syncingData = false
+    }
+
+    func didSelectCoupon(_ coupon: Coupon) {
+        selectedCoupon = coupon
+    }
+}
+
+// MARK: - Data for `MostActiveCouponsCardViewModel`
+//
+extension MostActiveCouponsCardViewModel {
+    var timeRangeText: String {
+        guard let timeRangeViewModel = StatsTimeRangeBarViewModel(timeRange: timeRange, timezone: siteTimezone) else {
```

Comment:
Nit: should we initialize `StatsTimeRangeBarViewModel` only after the time range is updated to avoid creating a new instance every time the view is rendered?

Reply:
Good suggestion. Done in https://github.com/woocommerce/woocommerce-ios/pull/12733/commits/603233993a13f134d96a3d5db7b4cb65a2818f68